src/include/stir/KOSMAPOSL/KOSMAPOSLReconstruction.h

```diff
@@ -0,0 +1,312 @@
+  virtual void update_estimate (TargetT& current_image_estimate);
+
+  int subiteration_counter;
+  double anatomical_sd;
+  mutable Array<3,float> distance;
+  /*! Create a matrix containing the norm of the difference between two feature vectors, \f$ \|  \boldsymbol{z}^{(n)}_j-\boldsymbol{z}^{(n)}_l \| \f$. */
+  /*! This is done for the PET image which keeps changing*/
+    void  calculate_norm_matrix(TargetT &normp,
+                                const int &dimf_row,
+                                const int &dimf_col,
+                                const TargetT& pet,
+                                Array<3,float> distance);
+
+  /*! Create a matrix similarly to calculate_norm_matrix() but this is done for the anatomical image, */
+  /*! which does not  change over iteration.*/
+    void  calculate_norm_const_matrix(TargetT &normm,
+                                const int &dimf_row,
+                                const int &dimf_col);
+
+  /*! Estimate the SD of the anatomical image to be used as normalisation for the feature vector */
+    void estimate_stand_dev_for_anatomical_image(double &SD);
+
+  // /*! Compute for each voxel, jl, of the PET image the linear combination between the coefficient \f$ \alpha_{jl} \f$ and the kernel matrix \f$ k_{jl} \f$\f$ */
+  // /*! The information is stored in the image, kImage */
+  // void full_compute_kernelised_image(TargetT& kernelised_image_out,
+  //                                    const TargetT& image_to_kernelise,
+  //                                    const TargetT& current_alpha_estimate);
+
+  //  /*! Similar to compute_kernelised_image() but this is the special case when the feature vectors contains only one non-zero element. */
+  //  /*! The computation becomes faster because we do not need to create norm matrixes*/
+  // void compact_compute_kernelised_image(TargetT& kernelised_image_out,
+  //                                       const TargetT& image_to_kernelise,
+  //                                       const TargetT& current_alpha_estimate);
+
+  /*! choose between compact_compute_kernelised_image() and  full_compute_kernelised_image()*/
+  void compute_kernelised_image(TargetT& kernelised_image_out,
+                                const TargetT& image_to_kernelise,
+                                const TargetT& current_alpha_estimate);
+
+  double calc_pet_kernel(int x, int y, int z,
+                         int min_x, int min_y, int min_z,
+                         int max_x, int max_y, int max_z,
+                         int dx, int dy, int dz,
+                         int min_dx, int min_dy, int min_dz,
+                         int max_dx, int max_dy, int max_dz,
+                         const TargetT& current_alpha_estimate,
+                         bool use_compact_implementation,
+                         const CartesianCoordinate3D<float>& grid_spacing,
+                         Array<3, float> distance);
```

Comment:
pass by ref?

Reply:
definitely, and `const&` presumably

Reply:
OK I change all the call to distance and it is already significantly faster. Almost there!